Residential Heat Pump Water Heater Improvements

nmerket opened the pull request:
- [x] Rename existing `WaterHeater:HeatPump` to `WaterHeater:HeatPumpPumpedCondenser`. 
 - [x] Create a new idf object for residential heat pump water heaters called `WaterHeater:HeatPump:WrappedCondenser`. The new object will be different in the following ways:
    - [x] `WaterHeater:HeatPump:WrappedCondenser` will only be compatible with a `WaterHeater:Stratified` tank (and `WaterHeater:HeatPump:PumpedCondenser` will only be compatible with a `WaterHeater:Mixed` tank)
    - [x] Inputs relating to condenser water pumping will be removed.
    - [x] Add inputs relating to the location of the condenser on the tank and change heat transfer calculation to allow for a distribution of heat to come in across several nodes.
    - [x] ~~Include warnings if the timestep is too long for reasonable results. Also, warn if not using subhourly draw profiles.~~
    - [x] Add an option to not allow the heat pump and element(s) to run simultaneously. This can be bool. The element(s) should be master (if both want to come on, the element gets priority).
    - [x] ~~Add an option to allow heating cycles to complete with the heat pump. This can also be a separate bool input. If this is true, once the element comes on, it stays on until the heating cycle is over instead of switching back to the heat pump once the tank has partially recovered.~~
 - [x] Change `WaterHeater:Stratified` IDD to allow the maximum number of nodes to be 12
 - [x] Change `WaterHeater:HeatPump:PumpedCondenser` and `WaterHeater:HeatPump:WrappedCondenser` to:
   - [x] Have a minimum air inlet temperature of -5 &deg;C
   - [x] Create new maximum air inlet temperature input and choose appropriate bounds. Default to 120F.
 - [x] ~~Add default/example performance curves (biquadratic curves for heating capacity and COP) for residential HPWHs. NREL has developed generic performance curves based on [lab testing of three manufacturer’s residential HPWHs](http://www.osti.gov/scitech/biblio/1025650). Add to `PerfCurves.idf`~~ don't make sense on their own without the HPWH model.
 - [x] Unit tests
 - [x] Integration tests
   - [x] Fix existing tests to reflect updated idd
 - [x] Documentation
 - [x] Talk to @Myoldmopar about how to work idd changes into the transition program.

nmerket commented:
Yes, I've been working on it, but haven't pushed my changes for a while.

Myoldmopar commented:
@nmerket Are you talking to a robot again? :wink:

nmerket commented:
Yes, a passive-agressive robot who�s implying I�m a slacker...

On Feb 11, 2015, at 8:27 AM, Edwin Lee <notifications@github.com<mailto:notifications@github.com>> wrote:


@nmerket<https://github.com/nmerket> Are you talking to a robot again? [:wink:]

�
Reply to this email directly or view it on GitHub<https://github.com/NREL/EnergyPlus/pull/4698#issuecomment-73899551>.

Myoldmopar commented:
AAAAAHHHHHHHHH!!!!!!!

mbadams5 commented:
It has achieved sentience!! @nrel-bot == Skynet

nmerket commented:
From @jmaguire1
> Also, this is the presentation I mentioned that Ben made about his HPWH model: http://www.aceee.org/files/pdf/conferences/hwf/2013/1A-larson.pdf. Slide 10 shows how he’s distributing heat from the condenser within the tank. I don’t really think it’s that important (at least for matching on an annual basis, it may matter some in getting better performance on a day by day basis), so I would say the even distribution you proposed is fine. We could always use TRNSYS to look at how important this is (may not be a bad idea before we finalize things, but I’d say it’s not something we need to do ASAP).
Something to think about.

nmerket commented:
## Updates
- The residential hpwh model works and a model can be built that matches lab testing. Still working on an idf for other lab results.
- Merged in the vshpwh (among other) changes from `develop`. Waiting for CI results to know which test idfs that I need to modify to be compatible with the new idd. 

## Next Steps
- Unit tests
- Integration test. (I'm thinking of using some of these lab results here.)
- Docs

We're almost there!

Myoldmopar commented:
Great, @nmerket.  This will be an awesome addition.  I can assist with figuring out idfs that need to change and with any transition rules.  Let me know if you want to chat.

nmerket commented:
@Myoldmopar Thanks. Do you have time sometime in the next few days to help me work out the IDF Transition from the old to the new and help me update the integration tests?

Myoldmopar commented:
@nmerket Any luck digging in further on this?  If not, come by after 2:30 and we can look at it.

nmerket commented:
I've been doing other things and haven't had a chance to get back into this. Are you going to be around tomorrow?

Myoldmopar commented:
Yeah, after lunch is fine (~1).

nmerket commented:
@Myoldmopar, I could use some help with the idf transition program. Here's what's different and will need to be transitioned:
- `Coil:WaterHeating:AirToWaterHeatPump` -> `Coil:WaterHeating:AirToWaterHeatPump:Pumped`. All fields stayed the same, just the name changed.
- `WaterHeater:Stratified` now can have 12 nodes. I updated all the ones associated with HPWHs.
  - If a Source Side Flow Control Mode (A21) or Indirect Alternate Setpoint Temperature Schedule Name (A22) are present, two empty input fields for nodes 11 and 12 Additional Loss Coefficient need to be inserted.
- `WaterHeater:HeatPump` -> `WaterHeater:HeatPump:PumpedCondenser`. I made all of these transitions in the testfiles directory manually. 
  - Remove Tank Object Type (A14)
  - Remove DX Coil Object Type (A18)
  - Remove Fan Object Type (A22)
  - Added field Maximum Inlet Air Temperature for Compressor Operation (N5). Leave blank.
  - Control Sensor Location In Stratified Tank (N8) is replaced with Control Sensor Height in Stratified Tank. Need to look up the appropriate control sensor height from the `WaterHeater:Stratified` object. If this field was omitted previously, it can be omitted again and it will default to `Heater1`.

nmerket commented:
@Myoldmopar did you see my comment about the transition program for residential HPWH? Will you do that on my branch before merging or after? Do you need any additional information from me?

mjwitte commented:
@nmerket Ideally, the transition changes should be added to this branch so they can go in at the same time.

Myoldmopar commented:
@nmerket I saw it, I was just ignoring it.  

OK, not really ignoring it, but waiting on a time to work on it.  

I've got transition changes to work on for @mjwitte as well. I'll try to get these things done today:
- [ ] Create a new branch `branch1` to bring up transition to the latest version, it should drop into develop pretty easily
- [ ] Create a branch based on `branch1` where I add in @mjwitte changes for transition 
- [ ] Merge `branch1` into this branch and add your changes.

nmerket commented:
I'll wait for your commits to this branch then. Otherwise, I think it's ready to go.

Myoldmopar commented on doc/src/docs/EngineeringReference/13f-EncyclopaedicRefs.md:
```diff
@@ -2172,7 +2172,55 @@ Under certain input parameters, the rating method will not succeed and a warning
+
+Each type of `WaterHeater:HeatPump:*` object is compatible with certain `Coil:WaterHeating:AirToWaterHeatPump:*` objects as shown in the following table:
+
+<table>
```
HTML tables should be class: table-striped to be uniform with all the rest of them.  Bonus points if you just change it in the css to apply to all tables.

Myoldmopar commented on src/EnergyPlus/DXCoils.hh:
```diff
@@ -1595,7 +1596,7 @@ namespace DXCoils {
 	//     NOTICE
 
-	//     Copyright � 1996-2014 The Board of Trustees of the University of Illinois
+	//     Copyright � 1996-2014 The Board of Trustees of the University of Illinois
```
Copyright symbol got botched...why does it say 2014, anyway :wink:?  You don't have to change this, I have a branch around here somewhere to update the year to 2015, and I'll update it to remove the symbol entirely.

rraustad replied:
Did you change the copyright symbol on purpose?

Myoldmopar commented:
@nmerket So can you summarize the transition changes in [this](https://github.com/NREL/EnergyPlusBuildSupport/blob/v8.4.0/release/Rules8-3-0-to-8-4-0.xls) file?  Make sure you update it on the v8.4.0 branch or you won't find that file.  You can check out that file from previous (v8.2.0, v8.3.0) branches for more examples.  If you need help, just let me know.

nmerket commented:
@Myoldmopar what do I do when I changed the name of an object, like `WaterHeater:HeatPump` to `WaterHeater:HeatPump:PumpedCondenser`?

Myoldmopar commented:
Hmm, changing an object name requires a Docker change, right @nllong?   :boat: 

OK ok ok.  Just put on there that you are changing the object name.  There is an example in the v8.3.0 [rules file](https://github.com/NREL/EnergyPlusBuildSupport/blob/v8.3.0/release/Rules8-2-0-to-8-3-0.xls).

rraustad commented on doc/src/docs/InputOutputReference/01b-InputOutputReference.md:
```diff
@@ -17369,16 +17360,13 @@ WaterHeater:HeatPump,
     ,                        !- Inlet Air Humidity Schedule Name
     ,                        !- Inlet Air Zone Name
-    WaterHeater:Mixed,       !- Tank Object Type
     HPWHPlantTank,           !- Tank Name
```
Same old question of requiring developers to provide type and name for objects. That requirement was probably never added to the developers guide since in a case like this the object type has only 1 choice.

mjwitte replied:
@nmerket The object type fields should remain.  The pair of fields for type and name ensure that the reference to another object will be unique.  I will comment further in the idd.

rraustad commented on doc/src/docs/InputOutputReference/01b-InputOutputReference.md:
```diff
@@ -17398,7 +17386,7 @@ NOTE: branch object required only when tank use inlet nodes are used.
     Central HPWH Branch,     !- Name
     0,                       !- Maximum Flow Rate {m3/s}
     ,                        !- Pressure Drop Curve Name
-    WaterHeater:HeatPump,    !- Component 1 Object Type
```
From this line down looks like a ZoneHVAC:EquipmentConnections object. The end of the HPWH object and beginning of the ZoneHVAC:EquipmentConnections object is missing here.

nmerket replied:
Some github weirdness is going on here. It's just in the diff display. It looks like it's skipping from 17372 to 17386 in the diff view without indicating that the skip is happening. When I open up the file that is committed here's what's there:

<hr/>

``` idf
WaterHeater:HeatPump:PumpedCondenser,
    PlantHeatPumpWaterHeater,!- Name
    PlantHPWHSch,            !- Availability Schedule Name
    PlantHPWHTempSch,        !- Compressor Setpoint Temperature Schedule Name
    2.0,                     !- Dead Band Temperature Difference {deltaC}
    HPPlantWaterInletNode,   !- Condenser Water Inlet Node Name
    HPPlantWaterOutletNode,  !- Condenser Water Outlet Node Name
    0.00115525,              !- Condenser Water Flow Rate {m3/s}
    1.00695,                 !- Evaporator Air Flow Rate {m3/s}
    OutdoorAirOnly,          !- Inlet Air Configuration
    ,                        !- Air Inlet Node Name
    ,                        !- Air Outlet Node Name
    HPPlantAirInletNode,     !- Outdoor Air Node Name
    HPPlantAirOutletNode,    !- Exhaust Air Node Name
    ,                        !- Inlet Air Temperature Schedule Name
    ,                        !- Inlet Air Humidity Schedule Name
    ,                        !- Inlet Air Zone Name
    HPWHPlantTank,           !- Tank Name
    HPWH Use Inlet Node,     !- Tank Use Side Inlet Node Name
    HPWH Use Outlet Node,    !- Tank Use Side Outlet Node Name
    HPWHPlantDXCoil,         !- DX Coil Name
    11.0,                    !- Minimum Inlet Air Temperature for Compressor Operation {C}
    Outdoors,                !- Compressor Location
    ,                        !- Compressor Ambient Temperature Schedule Name
    HPWHPlantFan,            !- Fan Name
    BlowThrough,             !- Fan Placement
    ,                        !- On Cycle Parasitic Electric Load {W}
    ,                        !- Off Cycle Parasitic Electric Load {W}
    ;                        !- Parasitic Heat Rejection Location
```

NOTE: branch object required only when tank use inlet nodes are used.

``` idf
  Branch,
    Central HPWH Branch,     !- Name
    0,                       !- Maximum Flow Rate {m3/s}
    ,                        !- Pressure Drop Curve Name
    WaterHeater:HeatPump:PumpedCondenser,    !- Component 1 Object Type
    PlantHeatPumpWaterHeater,!- Component 1 Name
    HPWH Use Inlet Node,     !- Component 1 Inlet Node Name
    HPWH Use Outlet Node,    !- Component 1 Outlet Node Name
    PASSIVE;                 !- Component 1 Branch Control Type




  WaterHeater:Mixed,
    HPWHPlantTank,           !- Name
    0.757,                   !- Tank Volume {m3}
    Plant Hot Water Setpoint Temp Schedule,  !- Setpoint Temperature Schedule Name
    2.0,                     !- Deadband Temperature Difference {deltaC}
    82.2222,                 !- Maximum Temperature Limit {C}
    CYCLE,                   !- Heater Control Type
    25000,                   !- Heater Maximum Capacity {W}
    0,                       !- Heater Minimum Capacity {W}
    ,                        !- Heater Ignition Minimum Flow Rate {m3/s}
    ,                        !- Heater Ignition Delay {s}
    ELECTRICITY,             !- Heater Fuel Type
    0.98,                    !- Heater Thermal Efficiency
    ,                        !- Part Load Factor Curve Name
    10,                      !- Off Cycle Parasitic Fuel Consumption Rate {W}
    ELECTRICITY,             !- Off Cycle Parasitic Fuel Type
    0,                       !- Off Cycle Parasitic Heat Fraction to Tank
    30,                      !- On Cycle Parasitic Fuel Consumption Rate {W}
    ELECTRICITY,             !- On Cycle Parasitic Fuel Type
    0,                       !- On Cycle Parasitic Heat Fraction to Tank
    Outdoors,                !- Ambient Temperature Indicator
    ,                        !- Ambient Temperature Schedule Name
    ,                        !- Ambient Temperature Zone Name
    HPWHPlantTank OA Node,   !- Ambient Temperature Outdoor Air Node Name
    0.0,                     !- Off Cycle Loss Coefficient to Ambient Temperature {W/K}
    0.0,                     !- Off Cycle Loss Fraction to Zone
    0.0,                     !- On Cycle Loss Coefficient to Ambient Temperature {W/K}
    0.0,                     !- On Cycle Loss Fraction to Zone
    ,                        !- Peak Use Flow Rate {m3/s}
    ,                        !- Use Flow Rate Fraction Schedule Name
    ,                        !- Cold Water Supply Temperature Schedule Name
    HPWH Use Inlet Node,     !- Use Side Inlet Node Name
    HPWH Use Outlet Node,    !- Use Side Outlet Node Name
    0.98,                    !- Use Side Effectiveness
    HPPlantWaterOutletNode,  !- Source Side Inlet Node Name
    HPPlantWaterInletNode,   !- Source Side Outlet Node Name
    0.98,                    !- Source Side Effectiveness
    autosize;                !- Use Side Design Flow Rate {m3/s}
```

rraustad replied:
You might have already done this. The NOTE line above should be indented to include in the block text.

rraustad commented on doc/src/docs/InputOutputReference/01b-InputOutputReference.md:
```diff
@@ -17523,8 +17511,340 @@ This output is the average part-load ratio of the heat pump water heater’s com
+
+This numeric field contains the dead band temperature difference in degrees Celsius. The heat pump compressor “cut-in” temperature is defined as the compressor set point temperature defined above minus this dead band temperature difference. The heat pump compressor cycles on when the water temperature in the tank falls below the “cut-in” temperature. The heat pump compressor remains on until the  ...
+
+In this model, the heat pump water heater’s DX compression system is considered the primary heat source and the water tank’s heater (element or burner) provides supplemental heat as necessary. Therefore, the cut-in temperature for the heat pump compressor (set point minus dead band temperature difference) is usually higher than the set point temperature for the heater (element or burner) in the a ...
```
Last sentence "...heat pump compressor is disabled...", disabled MAY be incorrect as the compressor could turn on if code permitted and the tank water temperature dropped sufficiently. Probably nothing to do here expect to note expected performance and possibly change this wording at a later date.

nmerket replied:
ok

rraustad commented on doc/src/docs/InputOutputReference/01b-InputOutputReference.md:
```diff
@@ -17523,8 +17511,340 @@ This output is the average part-load ratio of the heat pump water heater’s com
+
+#### Field: Evaporator Air Flow Rate
+
+This numeric field contains the air flow rate across the heat pump’s air coil (evaporator) in cubic meters per second. It is the actual air flow rate to be simulated, which may differ from the rated evaporator air volumetric flow rate specified for the heat pump’s DX coil (Ref. Coil:WaterHeating:AirToWaterHeatPump). Values must be greater than 0 or this field is autocalculatable. If autocalculate ...
```
3rd and 7th lines... (Ref. Coil:WaterHeating:AirToWaterHeatPump:Wrapped or ...AirToWaterHeatPump:\* as required).

rraustad commented on doc/src/docs/InputOutputReference/01b-InputOutputReference.md:
```diff
@@ -17523,8 +17511,340 @@ This output is the average part-load ratio of the heat pump water heater’s com
+
+#### Field: Tank Name
+
+This alpha field contains the name of the specific water heater tank (WaterHeater:Mixed object) used by this heat pump water heater. This must be a tank of type `WaterHeater:Stratified`.
```
(WaterHeater:Mixed object) needs to be changed to :Stratified, or Ref. WaterHeater:Stratified or deleted.

rraustad commented on doc/src/docs/InputOutputReference/01b-InputOutputReference.md:
```diff
@@ -17523,8 +17511,340 @@ This output is the average part-load ratio of the heat pump water heater’s com
+
+#### Field: Minimum Inlet Air Temperature for Compressor Operation
+
+This numeric field contains the minimum inlet air dry-bulb temperature entering the air coil (evaporator) and fan section, in degrees Celsius, below which the heat pump compressor does not operate. The minimum inlet air dry-bulb temperature should be greater than or equal to -5°C. If this field is left blank, the default value is 10°C.
```
I'd hope that -5C in not hard-coded in the model since users have asked to allow operation to lower temps. Changing the idd would be easy for users if GetInput did not fatal out. No action required here.

nmerket replied:
The -5C for minimum inlet air temperature for compressor operation is in the idd, but it also issues a warning [in the code here](https://github.com/NREL/EnergyPlus/blob/beef5746f3e6145bec0a155a8fcca1b5527b8f64/src/EnergyPlus/WaterThermalTanks.cc#L1547-L1552). I think it will still run, but the warning will show up in the .err file. Should that be removed then?

rraustad commented on doc/src/docs/InputOutputReference/01b-InputOutputReference.md:
```diff
@@ -17523,8 +17511,340 @@ This output is the average part-load ratio of the heat pump water heater’s com
+
+#### Field: Compressor Location
+
+This alpha (choice) field contains the location of the heat pump compressor and the air temperature for this location is used to control operation of the compressor’s crankcase heater in the Coil:WaterHeating:AirToWaterHeatPump object. Valid entries are **Schedule**, **Zone**, or **Outdoors**. If ‘Schedule’ is selected, a compressor ambient temperature schedule name must be defined in the field b ...
```
3rd line...Coil:WaterHeating:AirToWaterHeatPump:Wrapped

rraustad commented on doc/src/docs/InputOutputReference/01b-InputOutputReference.md:
```diff
@@ -17523,8 +17511,340 @@ This output is the average part-load ratio of the heat pump water heater’s com
+
+#### Water Heater Compressor Part Load Ratio
+
+This output is the average part-load ratio of the heat pump water heater’s compressor (as well as its water pump and fan) for the timestep being reported. This output is independent of the “Water Heater Part Load Ratio” (Ref. Water Heater Outputs) which represents the part- load ratio of the supplemental heater (element or burner) in the water tank. When the water tank’s (supplemental) heater set ...
```
3rd line...change "part-  load" to part-load

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -8392,9 +8392,221 @@ This output field is the average electricity consumption rate of the DX coil com
+
+#### Field: Rated Evaporator Air Flow Rate
+
+This numeric field defines the evaporator air volume flow rate in cubic meters per second at rated conditions. Values must be greater than 0. If this field is left blank or autocalculated   (field value = **autocalculate**), the default value is 5.035E-5 m<sup>3</sup>/s/W (31.25 cfm/MBH) multiplied by the Rated Heating Capacity specified above. When autocalculating the rated evaporator air volume ...
```
2nd line...spaces after autocalculated

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -8392,9 +8392,221 @@ This output field is the average electricity consumption rate of the DX coil com
+
+#### Field: Maximum Ambient Temperature for Crankcase Heater Operation
+
+This numeric field defines the maximum ambient temperature for crankcase heater operation in degree Celsius. The crankcase heater only operates when the air surrounding the compressor is below this maximum temperature value and the compressor is off   The ambient temperature surrounding the compressor is set by the Heat Pump:Water Heater parent object (field Compressor Location).
```
3rd line...compressor is off   The...and end of line...Heat Pump:Water Heater (no spaces)

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -8392,9 +8392,221 @@ This output field is the average electricity consumption rate of the DX coil com
+
+#### Field: Part Load Fraction Correlation Curve Name
+
+This alpha field defines the name of a **quadratic** or **cubic** performance curve (Ref: Performance Curves) that parameterizes the variation of electrical power input to the DX unit as a function of the part load ratio (PLR, sensible cooling load/steady-state sensible cooling capacity). The product of the rated EIR and EIR modifier curves is divided by the output of this curve to give the    ef ...
```
4th line... give the   effective   EIR... spacing

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -8392,9 +8392,221 @@ This output field is the average electricity consumption rate of the DX coil com
+
+           PLF = 1.0 + 0.0(PLR)
+
+Following is an example input for the `Coil:WaterHeating:AirToWaterHeatPump:Pumped` object:
```
Coil:WaterHeating:AirToWaterHeatPump:Wrapped

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -8392,9 +8392,221 @@ This output field is the average electricity consumption rate of the DX coil com
+
+#### Cooling Coil Crankcase Heater Electric Power[W]
+
+This output field is the average electricity consumption rate of the DX coil compressor   s crankcase heater in Watts for the timestep being reported. The crankcase heater operates only when the compressor is off and the air surrounding the compressor is below the Maximum Ambient Temperature for Crankcase Heater Operation, otherwise this output variable is set equal to 0.
```
1st line...compressor's

rraustad commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -8392,9 +8392,221 @@ This output field is the average electricity consumption rate of the DX coil com
+
+#### Cooling Coil Crankcase Heater Electric Energy [J]
+
+This output field is the total electricity consumption of the DX coil compressor   s crankcase heater in Joules for the timestep being reported. This output is also added to a meter with Resource Type = Electricity, End Use Key = DHW, Group Key = Plant (ref. Output:Meter objects).
```
1st line...compressor's

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -47098,6 +47116,165 @@ Coil:WaterHeating:AirToWaterHeatPump,
+       \memo Heat pump water heater (HPWH) heating coil, air-to-water direct-expansion (DX)
+       \memo system which includes a water heating coil, evaporator air coil, evaporator
+       \memo fan, electric compressor, and water pump. Part of a WaterHeater:HeatPump:PumpedCondenser system.
+       \min-fields 14
```
Part of a WaterHeater:HeatPump:WrappedCondenser

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -47098,6 +47116,165 @@ Coil:WaterHeating:AirToWaterHeatPump,
+Coil:WaterHeating:AirToWaterHeatPump:Wrapped,
+       \memo Heat pump water heater (HPWH) heating coil, air-to-water direct-expansion (DX)
+       \memo system which includes a water heating coil, evaporator air coil, evaporator
+       \memo fan, electric compressor, and water pump. Part of a WaterHeater:HeatPump:PumpedCondenser system.
```
Part of a WaterHeater:HeatPump:WrappedCondenser

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -47098,6 +47116,165 @@ Coil:WaterHeating:AirToWaterHeatPump,
+       \note The compressor crankcase heater only operates when the dry-bulb temperature of air
+       \note surrounding the compressor is below the Maximum Ambient Temperature for Crankcase
+       \note Heater Operation and the DX coil is off.  The ambient temperature surrounding the
+       \note compressor is set by the WaterHeater:HeatPump:PumpedCondenser parent object (field Compressor Location).
```
WaterHeater:HeatPump:WrappedCondenser

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -47098,6 +47116,165 @@ Coil:WaterHeating:AirToWaterHeatPump,
+       \note The compressor crankcase heater only operates when the dry-bulb temperature of air
+       \note surrounding the compressor is below the Maximum Outdoor Temperature for Crankcase
+       \note Heater Operation and the unit is off. The ambient temperature surrounding the
+       \note compressor is set by the WaterHeater:HeatPump:PumpedCondenser parent object (field Compressor Location).
```
WaterHeater:HeatPump:WrappedCondenser

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -64280,7 +64457,7 @@ WaterHeater:Mixed,
        \memo Water heater with stratified, multi-node water tank. May be used to model a tankless
        \memo water heater (small tank volume), a hot water storage tank (zero heater capacity), or
-       \memo a heat pump water heater (see WaterHeater:HeatPump.)
+       \memo a heat pump water heater (see WaterHeater:HeatPump:*.)
```
Can only be used with WaterHeater:HeatPump:WrappedCondenser ? I guess :\* is OK

nmerket replied:
The `WaterHeater:Stratified` tank model can be used with either of the `WaterHeater:HeatPump:*` objects. However, the `WaterHeater:HeatPump:WrappedCondenser` can only be used with the `WaterHeater:Stratified` tank model. I tried to outline the compatibility of the models [here](https://github.com/NREL/EnergyPlus/blob/hpwh_residential/doc/src/docs/EngineeringReference/13f-EncyclopaedicRefs.md#overview)

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -64868,37 +65045,257 @@ WaterHeater:HeatPump,
+       \note A schedule value of 0 denotes inlet air is drawn only from the zone.
+       \note A schedule value of 1 denotes inlet air is drawn only from outdoors.
+       \note Schedule values between 0 and 1 denote a mixture of zone and outdoor air
+       \note proportional to the schedule value.
```
suggestion...schedule value (i.e., 0.4 = 40% outdoor air, 60% zone air).

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -64868,37 +65045,257 @@ WaterHeater:HeatPump,
+       \note Simultaneous (default) means that both the tank heating element and
+       \note heat pump are used at the same time recover the tank temperature.
+  N8,  \field Control Sensor 1 Height In Stratified Tank
+       \note Used to indicate height of control sensor if Tank Object Type is WaterHeater:Stratified
```
suggestion...control sensor for Tank Object Type = WaterHeater:Stratified

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -64868,37 +65045,257 @@ WaterHeater:HeatPump,
+       \maximum 1.0
+       \default 1.0
+  N10; \field Control Sensor 2 Height In Stratified Tank
+       \note Used to indicate height of control sensor if Tank Object Type is WaterHeater:Stratified
```
suggestion...control sensor for Tank Object Type = WaterHeater:Stratified

rraustad commented on src/EnergyPlus/DataHVACGlobals.cc:
```diff
@@ -128,14 +128,15 @@ namespace DataHVACGlobals {
 	int const CoilDX_CoolingTwoStageWHumControl( 5 );
-	int const CoilDX_HeatPumpWaterHeater( 6 );
+	int const CoilDX_HeatPumpWaterHeaterPumped( 6 );
+	int const CoilDX_HeatPumpWaterHeaterWrapped( 31 );
```
Why didn't you just change the numbering 1-n ? It doesn't matter if CoilDX_HeatPumpWaterHeaterWrapped is now 7, but it does matter of a future developer misses the indexing. I guess the compiler will catch it?

rraustad commented on src/EnergyPlus/DataPlant.cc:
```diff
@@ -177,7 +177,8 @@ namespace DataPlant {
 	int const TypeOf_Generator_FCExhaust( 15 );
-	int const TypeOf_HeatPumpWtrHeater( 16 );
+	int const TypeOf_HeatPumpWtrHeaterPumped( 16 );
+	int const TypeOf_HeatPumpWtrHeaterWrapped( 92 );
```
same comment, 1-n

rraustad commented on src/EnergyPlus/DataZoneEquipment.cc:
```diff
@@ -553,9 +553,12 @@ namespace DataZoneEquipment {
-
+					
+					} else if ( SELECT_CASE_var == "WATERHEATER:HEATPUMP:WRAPPEDCONDENSER" ) {
+						ZoneEquipList( ControlledZoneNum ).EquipType_Num( ZoneEquipTypeNum ) = HPWaterHeater_Num;
```
I thought these had unique indexes? (i.e., both use HPWaterHeater_Num), maybe that's just the TypeOf as shown below. 

This one is OK, they are both HPWHs.

rraustad commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -5375,9 +5376,9 @@ namespace VariableSpeedCoils {
 			WhichCoil = FindItemInList( CoilName, VarSpeedCoil.Name(), NumWatertoAirHPs );
 			if ( WhichCoil != 0 ) {
-				if ( CoilType == "COIL:HEATING:WATERTOAIRHEATPUMP:VARIABLESPEEDEQUATIONFIT" || CoilType == "COIL:HEATING:DX:VARIABLESPEED" ) {
+				if ( SameString( CoilType, "COIL:HEATING:WATERTOAIRHEATPUMP:VARIABLESPEEDEQUATIONFIT" ) || SameString( CoilType, "COIL:HEATING:DX:VARIABLESPEED" ) ) {
```
Doesn't switching to SameString add clock cycles?

nmerket replied:
Yes, but everywhere else in this function was using `SameString` for similar comparisons. I took it to mean that we needed to do a case-insensitive comparison and that it was just overlooked.

rraustad replied:
OK, it's probably needed since users could enter lower case for type.

rraustad commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -2891,7 +2891,7 @@ namespace ZoneEquipmentManager {
 					SimStandAloneERV( PrioritySimOrder( EquipTypeNum ).EquipName, ActualZoneNum, FirstHVACIteration, SysOutputProvided, LatOutputProvided, ZoneEquipList( ControlledZoneNum ).EquipIndex( EquipPtr ) );
 
-				} else if ( SELECT_CASE_var == HPWaterHeater_Num ) { // 'WaterHeater:HeatPump'
+				} else if ( SELECT_CASE_var == HPWaterHeater_Num ) { // 'WaterHeater:HeatPump:PumpedCondenser'
```
So here, it appears HPWaterHeater_Num refers to a pumped water heater, not a wrapped WH. See line 560 above.

rraustad commented:
Very thorough. I would suggest running a test file with both pumped and wrapped versions of HPWH to test compatibility. Approved after comments are addressed.

rraustad commented:
Lastly, what are the numerical diff's in the test suite files? I would expect no diff's in non-HPWH files and no (or very very small) diff's in pumped HPWH files.

rraustad commented:
Example file HeatPumpWaterHeaterStratified has this type warning:
*\* Warning *\* WaterHeater:Stratified = HPWHTANK:  More Additional Loss Coefficients were entered than the number of nodes; extra coefficients will not be used

These should be eliminated.

I also see eio diff's in warmup convergence. I assume that also means diff's in the results. How different would be the question.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -64816,50 +65000,43 @@ WaterHeater:HeatPump,
        \object-list ZoneNames
        \note Used only if Inlet Air Configuration is ZoneAirOnly or ZoneAndOutdoorAir.
        \note Otherwise, leave field blank.
-  A14, \field Tank Object Type
```
@nmerket Object Type fields should remain.  Curious what the motivation was to remove these?  
@Myoldmopar How can we document this intent to avoid future redos?

nmerket replied:
@mjwitte The motivation was that they are redundant since the object type field is on the object itself. I suppose I hadn't considered the possibility that EnergyPlus would support a user specifying a `WaterHeater:Mixed` and `WaterHeater:Stratified` object with the same name. Does EnergyPlus really need to support that?

@Myoldmopar Changing this will change the transition rules. I don't think any other changes noted on this pull request will.

Myoldmopar replied:
@nmerket OK, Let me know when you update the spreadsheet.

mjwitte replied:
@nmerket @Myoldmopar I see two main concerns here.
1.  Uniqueness of names - It's not that we _need_ to support the same name on two water heaters, but rather do we enforce unique names across all water heaters?  Currently, the only unique name enforcement that is universal is that all object names must be unique within a specific object-type.  There are places that enforce unique names across a group of similar objects (e.g. schedules).  I don't know about the waterheaters - would have to look at the code.
2. Consistency of input - The last time this was discussed (years ago), the development team consensus was to use the dual-field type+name approach when referencing other objects in HVAC system connections.  Clearly, we aren't consistent with this everywhere - we don't have object-type fields when referencing a schedule name.  At it's possible that a user could name a schedule the same as some other object.  EnergyPlus knows it's looking for a schedule.  The IDD \object-list tags are the only way to inform an interface that this field is a schedule name.
   If we want to adopt a new "rule" for this, then let's have a discussion that includes the OpenStudio team and bounce it off other interface developers.  Maybe they don't care which way it's done, but I would expect that they want consistency.  @kbenne ?

nmerket replied:
@mjwitte I see your points here. Especially the consistency one. I will change it back to require object type in all these places. Sigh. 

For the future, we really should have some developer documentation for object design. There were several things that would have saved me time with here.

mjwitte replied:
@nmerket Sorry about that.  Thanks for putting the field back in.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -64868,37 +65045,257 @@ WaterHeater:HeatPump,
+  A12, \field Tank Name
+       \required-field
+       \type object-list
+       \object-list WaterHeaterNames
```
@nmerket If this is limited to only stratified water heaters, then a new reference is needed that only includes those.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -64816,50 +65000,43 @@ WaterHeater:HeatPump,
        \type object-list
-       \object-list HeatPumpWaterHeaterDXCoils
+       \object-list HeatPumpWaterHeaterDXCoilsPumped
 	   \object-list HeatPumpWaterHeaterDXCoilsVariableSpeed
```
@nmerket This line has a stray tab in it.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -64711,12 +64896,11 @@ WaterHeater:Sizing,
-       \min-fields 31
+WaterHeater:HeatPump:PumpedCondenser,
+       \min-fields 29
        \memo This object models an air-source heat pump for water heating.
```
@nmerket The memos for these two waterheater:heatpump objects should have a sentence explaining the difference.  Hindsight (not suggesting we change this now) I only see about three fields that are different between these two objects, so we could have kept this a WaterHeater:HeatPump and added a few new fields.  Also, the condenser inlet/outlet node fields could be eliminated and mined from the coil object (also not suggesting you change that now).
@Myoldmopar At the risk of opening a huge debate, I'm going to post a new issue to discuss and clarify object design conventions and make sure we get that documented.  See #4995

Myoldmopar commented:
@nmerket I pulled in the transition changes for 8.3 to 8.4.  I then hijacked this branch to update some of the packaging rules since I forgot to do that over in develop.  

Anyway, now what I need from you is to define the IDD change rules over in BuildSupport.  Point your browser to keyword `https://github.com/NREL/EnergyPlusBuildSupport/`, and check out that repo locally. Switch to branch v8.4.0, browse to release/ and update the Rules8-3-0-to-8-4-0.xls file.  If you need clarification of how to fill out that sheet, just let me know.  Once the transition rules are clearly defined and nailed down, I'll then update the /src/Transition source code here in your branch, do some testing, and we'll be ready to go.

nmerket commented:
Just got back from vacation and it looks like I have a few things to address here. I'll start digging through all your comments and get back in the next few days. I just wanted to let you all know I'm not ignoring you.

Myoldmopar commented:
@nmerket Are the transition rules easy enough to describe right here?  I can fill in the rules sheet if so.  Example:
- Object "Foo", name change to "Foo2"
  - 8.3 field "bar", which is index {F5, N2} is removed
  - 8.4 new field "bar2" will be inserted at index {F8, N4}.  The value will be `1/bar`.

I'd like to wrap the transition for this up and help get this all done real soon for my own selfish reasons.

nmerket commented:
@Myoldmopar I thought I already updated those rules [here](https://github.com/NREL/EnergyPlusBuildSupport/blob/v8.4.0/release/Rules8-3-0-to-8-4-0.xls).

Also, sorry for not getting to this. I've had other fires to put out this week and should be able to focus on this next week.

Myoldmopar commented:
@nmerket OK, thanks.  That slipped by.  I'll update transition accordingly and push to this branch.

nmerket commented:
@Myoldmopar And this whole time I thought nothing slipped by you.

Myoldmopar commented:
I'm allowed 2 per year.  I'm right on schedule.

OK I'll get the code updated this afternoon likely and pushed up here.  Once that is complete, is this ready for any final reviews and merge?  Or do you still have something?

Myoldmopar commented:
@nmerket I went ahead and pushed up the transition changes.  The last commit (beefb5c) even includes support for the anticipated Tank Type field being added back into the IDD.  The transition is working well and matching the changes you made manually in the idfs.  Once you have the IDD, IDFs, and source updated for the tank type field, I'll re-test Transition to verify it works.  I have a feeling this branch will also become conflicted after some other transition work gets merged into develop, but I'll resolve that.

Myoldmopar commented:
@nmerket OK, I merged develop in, which pulled in the basis of the transition work.  Now if you can do the update for the re-added TYPE field, I'll then re-test the transition source, and this should be good to go.

nmerket commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -6751,7 +6752,7 @@ namespace VariableSpeedCoils {
 	//     NOTICE
 
-	//     Copyright � 1996-2014 The Board of Trustees of the University of Illinois
+	//     Copyright � 1996-2014 The Board of Trustees of the University of Illinois
```
@Myoldmopar every time I open this in xcode and save it, it changes the the &copy; to some unknown character. Not sure how to stop it from doing that.

Myoldmopar replied:
By me taking out the symbol entirely...please hold and I'll probably submit a PR to fix the copyright things...

Myoldmopar replied:
@nmerket As soon as #5035 merges in, pull develop into this branch and resolve the conflict in the copyright line by just choosing the new plain text version and we'll never have to worry about these again.

mjwitte commented:
@nmerket @Myoldmopar In case you haven't noticed, this branch is over 100 commits behind develop, so one of you will need to merge this up.

nmerket commented:
Well, I made the changes that should address everything here. Then I merged in develop. My HPWH tests seem fine, but now some totally unrelated unit tests are failing:

```
[----------] 20 tests from InputProcessorFixture
[ RUN      ] InputProcessorFixture.compareIDF
/Users/nmerket/energyplus/gitrepo/tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc:402: Failure
Value of: IDFRecords( index ).ObjectDefPtr
  Actual: 741
Expected: object_def_ptr
Which is: 739
Segmentation fault: 11
```

@Myoldmopar any idea what's going on? Did I totally botch the merge?

mbadams5 commented:
@nmerket No, you didn't botch anything. You added two new objects to Energy+.idd. So the test is expecting Output:SQLite at 739, but it is 741 now. You will need to change this (and maybe a couple other tests) to expect 741 now.

I am surprised though that it seg faults. I may try to look into that to beef it up so the seg fault doesn't happen again.

nmerket commented:
@mbadams5 Should I wait until #5036 is merged into develop and then pull them into this branch?

Myoldmopar commented:
@nmerket OK, if this passes the IP unit tests, are you done with it?  I'll go back and re-check the transition code changes I made, but I want to wait until it is all ready otherwise.

nmerket commented:
@Myoldmopar I really, really hope I'm done with it. I'd like to see the CI results before declaring it so.

Myoldmopar commented:
@nmerket Look at the area around this comment [here](https://github.com/NREL/EnergyPlus/pull/4698/#commits-pushed-25f49e4).  Note that above these comments, there are only a handful of failures/diffs.  After this commit, the number of failures goes way up.  Any reason? Intentional?

nmerket commented:
@Myoldmopar Yes, I have noticed that. I was hoping it was because we got so far behind `develop`, but it doesn't appear that is the case. The errors seem to be all over the place in things I didn't even touch. I'm wondering if it was the two renumbering commits de9532350e87d577f6ebf8c27ed75848cfc1e2af and 20cb70807a8762529f4f4a671acdcd670a3dc266 that did it. I made them in response to a [comment](https://github.com/NREL/EnergyPlus/pull/4698#discussion_r33288555) by @rraustad. I previously didn't mess with those because I was afraid of domino effects. I'll revert those two commits and see if that fixes it.

rraustad commented:
Renumbering should not be the cause (I do it all the time).

nmerket commented:
@rraustad Sorry, I didn't intend to blame you for it. I'm sure there's a right way to renumber and then what I did. I'll take full credit for that. Anyway, when I reverted those two commits the problems disappeared in the integration tests. I just pushed it back to this branch. We'll see if it makes a difference in the regression tests as well. I have a suspicion it will.

@mbadams5 Thanks for the updated IDD unit testing code.

Myoldmopar commented:
@nmerket @rraustad So it looks like the re-numbering did help, which is expected in this case because it was pretty clear that caused the diffs, but overall a bit surprising.  Screams of a hardwired number somewhere in the code to me...  

So with that fixed, I'll try to tackle the final transition code today and I'll pull develop in at the same time, so long as it isn't a deep conflict.

nmerket commented:
Looking at the regression test results, they all seem expected and minor at this point, so yay!

@mbadams5  merged `develop` into this branch yesterday afternoon, so hopefully it won't be too far off. 

We're so close on this!
